rpc: include_unsafe option for fundrawtransaction

t-bast opened the pull request:
Allow RPC users to opt-in to unsafe inputs when funding a raw transaction.

Applications that need to manage a complex RBF flow (such as lightning nodes using anchor outputs) are very limited if they can only use safe inputs.

I also added this option to `send` and `walletcreatefundedpsbt` who internally delegate to `fundrawtransaction`.

Fixes #21299

luke-jr commented on src/wallet/wallet.cpp:
```diff
@@ -2613,7 +2614,7 @@ SigningResult CWallet::SignMessage(const std::string& message, const PKHash& pkh
 }
 
-bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, bool lockUnspents, const std::set<int>& setSubtractFeeFromOutputs, CCoinControl coinControl)
+bool CWallet::FundTransaction(CMutableTransaction& tx, CAmount& nFeeRet, int& nChangePosInOut, bilingual_str& error, bool includeUnsafe, bool lockUnspents, const std::set<int>& setSubtractFeeFromOutputs, CCoinControl coinControl)
```
I suggest adding `includeUnsafe` into `CCoinControl` instead.

t-bast replied:
I completely agree, I mentioned that in the PR description. If there's a concept ACK, `CCoinControl` would be a good place to store that bool. Do you prefer `includeUnsafe` (more consistent with the RPC) or `fSafeOnly` (more consistent with the inner codebase)?

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/0dc8808230160e2da2b2f5aa943f5a3bac7dcb60

t-bast committed:
rpc: include_unsafe option for fundrawtransaction

Allow RPC users to opt-in to unsafe inputs when funding a raw transaction.

Applications that need to manage a complex RBF flow (such as lightning
nodes using anchor outputs) are very limited if they can only use safe inputs.

Fixes #21299

t-bast commented on src/wallet/wallet.cpp:
```diff
@@ -2797,7 +2798,7 @@ bool CWallet::CreateTransactionInternal(
         {
             std::vector<COutput> vAvailableCoins;
-            AvailableCoins(vAvailableCoins, true, &coin_control, 1, MAX_MONEY, MAX_MONEY, 0);
+            AvailableCoins(vAvailableCoins, !coin_control.m_include_unsafe_inputs, &coin_control, 1, MAX_MONEY, MAX_MONEY, 0);
```
We could remove the `fOnlySafe` argument to `AvailableCoins` now that this is provided by `coin_control`.

The only gotcha is that `AvailableCoins` can be called without a `coin_control`, but I checked the codebase and in all the places where that happens `fOnlySafe` is set to `true`, so we could set it inside `AvailableCoins` to `const bool only_safe = {coinControl ? !coinControl->m_include_unsafe_inputs : true};`.

Do you want me to apply that change in this PR?

t-bast commented:
@achow101 @luke-jr would you have some time to give a quick pass to this hopefully very simple PR and let me know if there's a concept ACK and it's worth pursuing? :pray: 

I know, Taproot is happening (is it?), I understand if you don't have much time available :smile:

achow101 commented:
ACK 16076bb2a8517c6a8b101ad04e220ad3b3f71bcf

t-bast commented:
Is there something I can do to help this PR move forward?

We really need it to unblock anchor outputs in lightning (we can run a modified `bitcoind` with this patch applied for our own node, but we can't ask our users to run a non-release `bitcoind` version).

renepickhardt commented:
Looks good to me.

Could there go anything other wrong than a transaction becoming invalid because an input disappeared? (this of course would not be really wrong)

t-bast commented:
> Could there go anything other wrong than a transaction becoming invalid because an input disappeared?

It's a very good question, it's part of my "unknown unknowns" and why I'm curious to get feedback from more people here.
It seems to me that there isn't anything else, but there are many things I don't know so I wouldn't rely on my instincts only ¯\_(ツ)_/¯

jonatack commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -928,6 +929,29 @@ def test_transaction_too_large(self):
         self.nodes[0].generate(10)
         assert_raises_rpc_error(-4, "Transaction too large", recipient.fundrawtransaction, rawtx)
 
+    def test_include_unsafe(self):
```
As `FundTransaction()` has other callers like the `walletcreatefundedpsbt` and `send` RPCs, might be a good idea to have a similar `test_include_unsafe` tests that check for present/future regressions in `rpc_test_psbt.py` and `wallet_send.py`.

t-bast replied:
Good idea, will do.

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/c3eb1994c24167bedb114e34b53dc7cbeb9ff403
I also added explicit support for these APIs, it makes more sense that way.

jonatack commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3203,6 +3208,9 @@ static RPCHelpMan fundrawtransaction()
                         {
                             {"add_inputs", RPCArg::Type::BOOL, /* default */ "true", "For a transaction with existing inputs, automatically include more if they are not enough."},
+                            {"include_unsafe", RPCArg::Type::BOOL, /* default */ "false", "Include inputs that are not safe to spend.\n"
+                                                          "Beware that the resulting transaction may become invalid if one of the unsafe inputs disappears.\n"
```
s/Beware/Be aware/ or s/Beware that/Warning:/

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/c3eb1994c24167bedb114e34b53dc7cbeb9ff403

jonatack commented:
This would need a release note.

t-bast commented:
> This would need a release note.

Added in https://github.com/bitcoin/bitcoin/pull/21359/commits/c3eb1994c24167bedb114e34b53dc7cbeb9ff403, let me know if that's how it should be done.

t-bast commented on test/functional/wallet_send.py:
```diff
@@ -133,6 +139,11 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
+        if include_unsafe:
+            from_balance = from_wallet.getbalances()["mine"]["untrusted_pending"]
+        else:
+            from_balance = from_wallet.getbalance()
```
I'm not sure that change should be done in this PR.
It may be worth sweeping all uses of `getbalance()` to replace them with `getbalances()["mine"]["trusted"]` in a separate PR, don't you think?
I volunteer to make that follow-up PR right after this one if that sounds good to you.

instagibbs replied:
it's new code, if you don't mind conforming to the newer standard for new code makes any future PR to "sweep" it up smaller?

t-bast replied:
Fair enough, and this is quite simple, done in https://github.com/bitcoin/bitcoin/pull/21359/commits/1bb252669dc5e8feb41e4975d609108eb28c8667

instagibbs commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3203,6 +3208,9 @@ static RPCHelpMan fundrawtransaction()
                     {"options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "for backward compatibility: passing in a true instead of an object will result in {\"includeWatching\":true}",
                         {
                             {"add_inputs", RPCArg::Type::BOOL, /* default */ "true", "For a transaction with existing inputs, automatically include more if they are not enough."},
+                            {"include_unsafe", RPCArg::Type::BOOL, /* default */ "false", "Include inputs that are not safe to spend.\n"
```
Might want to explain what is unsafe about them speifically, otherwise it seems to lead to the "why are you allowing it" question.

t-bast replied:
Makes sense, anything that can improve documentation is worth adding.
I copied the explanation from the `safe` flag in https://github.com/bitcoin/bitcoin/pull/21359/commits/1bb252669dc5e8feb41e4975d609108eb28c8667, let me know if you think the wording should be updated to be clearer.

jonatack commented on test/functional/rpc_psbt.py:
```diff
@@ -394,6 +394,14 @@ def run_test(self):
+        self.nodes[0].sendtoaddress(wunsafe.getnewaddress(), 2)
+        self.sync_mempools()
+        assert_raises_rpc_error(-4, "Insufficient funds", wunsafe.walletcreatefundedpsbt, [], [{self.nodes[0].getnewaddress():1}])
+        wunsafe.walletcreatefundedpsbt([], [{self.nodes[0].getnewaddress():1}], 0, {"include_unsafe": True})
```
style nit if you need to retouch (same on previous line)
```suggestion
        wunsafe.walletcreatefundedpsbt([], [{self.nodes[0].getnewaddress(): 1}], 0, {"include_unsafe": True})
```

t-bast replied:
The rest of that file doesn't put a whitespace in that case, this is why I used this format to stay coherent across the file, but if you think it makes sense to fix I should change it everywhere in the file, right?

jonatack replied:
I wouldn't worry about it apart from the new code and even then it's a pico nit.

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/2bf0b1e7ee28dbc06aebf832c0480d257dad7294

jonatack commented on test/functional/wallet_send.py:
```diff
@@ -33,12 +33,16 @@ def skip_test_if_missing_module(self):
+        if include_unsafe:
+            from_balance_before = from_wallet.getbalances()["mine"]["untrusted_pending"]
+        else:
+            from_balance_before = from_wallet.getbalances()["mine"]["trusted"]
```
While this works for the purposes of the test, should `from_balance` include "trusted" as well:
```python
        from_balance_before = from_wallet.getbalances()["mine"]["trusted"]
        if include_unsafe:
            from_balance_before += from_wallet.getbalances()["mine"]["untrusted_pending"]
```

t-bast replied:
That makes more sense indeed, done in https://github.com/bitcoin/bitcoin/pull/21359/commits/0709545365d045b24563b4683881f65a2b34afdd

jonatack commented on test/functional/wallet_send.py:
```diff
@@ -133,6 +139,11 @@ def test_send(self, from_wallet, to_wallet=None, amount=None, data=None,
+        if include_unsafe:
+            from_balance = from_wallet.getbalances()["mine"]["untrusted_pending"]
+        else:
+            from_balance = from_wallet.getbalances()["mine"]["trusted"]
```
same here
```python
        from_balance = from_wallet.getbalances()["mine"]["trusted"]
        if include_unsafe:
            from_balance += from_wallet.getbalances()["mine"]["untrusted_pending"]
```

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/0709545365d045b24563b4683881f65a2b34afdd

jonatack reviewed (commented):
ACK 4e4d7c01b4134ea3b920345ee2a62e2b06cd458e

Some comments below.

jonatack commented:
ACK 0709545365d045b24563b4683881f65a2b34afdd

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -2535,6 +2534,13 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+            // received from other wallets.
+            if (coin_control.m_include_unsafe_inputs && SelectCoinsMinConf(value_to_select,
+                                                        CoinEligibilityFilter(0, 0, max_ancestors-1, max_descendants-1, true /* include_partial_groups */),
+                                                        vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) {
```
nit, I think it's best to not indent arguments the same as the function they are a part of. Maybe something like
```diff
-            if (coin_control.m_include_unsafe_inputs && SelectCoinsMinConf(value_to_select,
-                                                        CoinEligibilityFilter(0, 0, max_ancestors-1, max_descendants-1, true /* include_partial_groups */),
-                                                        vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) {
+            if (coin_control.m_include_unsafe_inputs
+                && SelectCoinsMinConf(
+                    value_to_select,
+                    CoinEligibilityFilter(0, 0, max_ancestors-1, max_descendants-1, true /* include_partial_groups */),
+                    vCoins, setCoinsRet, nValueRet, coin_selection_params, bnb_used)) {
                 return true;
```

t-bast replied:
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/2bf0b1e7ee28dbc06aebf832c0480d257dad7294

jonatack reviewed (commented):
ACK a5c3d9cc4cafef3434c97656ab0d4b4fe4727d51

Happy to re-ack for https://github.com/bitcoin/bitcoin/pull/21359#discussion_r620213267 and https://github.com/bitcoin/bitcoin/pull/21359#discussion_r623679159.

jonatack commented on src/wallet/wallet.cpp:
```diff
@@ -2535,6 +2534,14 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
             }
+            // Try with unsafe inputs if they are allowed. This may spend unconfirmed outputs
+            // received from other wallets.
+            if (coin_control.m_include_unsafe_inputs 
```
linter appeasement time (sorry about that if it's from my suggestion)
```suggestion
            if (coin_control.m_include_unsafe_inputs
```

t-bast replied:
My bad! Fixed in https://github.com/bitcoin/bitcoin/pull/21359/commits/ce70f322ceb79fe8449d20967c810efcae4881f0

jonatack commented:
re-ACK 1b1dad3efbbdf86f5426bf9c43902f40cb9a06c0

ariard commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4030,6 +4038,9 @@ static RPCHelpMan send()
             {"options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "",
                 {
                     {"add_inputs", RPCArg::Type::BOOL, RPCArg::Default{false}, "If inputs are specified, automatically include more if they are not enough."},
+                    {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
```
Did we already consider for spending replacement transactions from our wallets before this PR ? `isRBFOptIn`doesn't seem used in `FundTransaction` path. Unsafe inputs definition doesn't seem to be tied to replaceability only key externality ?

t-bast commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3205,6 +3210,9 @@ static RPCHelpMan fundrawtransaction()
                     {"options", RPCArg::Type::OBJ, RPCArg::Optional::OMITTED_NAMED_ARG, "for backward compatibility: passing in a true instead of an object will result in {\"includeWatching\":true}",
                         {
                             {"add_inputs", RPCArg::Type::BOOL, RPCArg::Default{true}, "For a transaction with existing inputs, automatically include more if they are not enough."},
+                            {"include_unsafe", RPCArg::Type::BOOL, RPCArg::Default{false}, "Include inputs that are not safe to spend (unconfirmed transactions from outside keys and unconfirmed replacement transactions).\n"
```
It took the definition of `fSafe` here: https://github.com/bitcoin/bitcoin/blob/2b45cf0bcdb3d2c1de46899e30885c953b57b475/src/wallet/wallet.h#L595
Is there consensus on updating this in every place where it's used?

t-bast commented on src/wallet/wallet.cpp:
```diff
@@ -2535,6 +2534,14 @@ bool CWallet::SelectCoins(const std::vector<COutput>& vAvailableCoins, const CAm
+            // received from other wallets.
+            if (coin_control.m_include_unsafe_inputs
+                && SelectCoinsMinConf(
+                    value_to_select, CoinEligibilityFilter(0, 0, max_ancestors-1, max_descendants-1, true /* include_partial_groups */),
```
Fixed in https://github.com/bitcoin/bitcoin/pull/21359/commits/11d6459b6e101f05f36e13799c400bef82d2fc21

t-bast commented on test/functional/rpc_fundrawtransaction.py:
```diff
@@ -928,6 +929,29 @@ def test_transaction_too_large(self):
+        tx_dec = wallet.decoderawtransaction(fundedtx['hex'])
+        assert any([txin['txid'] == txid and txin['vout'] == vout for txin in tx_dec['vin']])
+        signedtx = wallet.signrawtransactionwithwallet(fundedtx['hex'])
+        wallet.sendrawtransaction(signedtx['hex'])
```
Done in https://github.com/bitcoin/bitcoin/pull/21359/commits/11d6459b6e101f05f36e13799c400bef82d2fc21
